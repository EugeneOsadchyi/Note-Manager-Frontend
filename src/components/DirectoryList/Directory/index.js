import React from 'react';
import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faFolder, faFolderOpen } from '@fortawesome/free-solid-svg-icons';


const Directory = ({
  name,
  opened,
  folderClicked,
  doubleClicked,
}) => {
  const icon = opened ? faFolderOpen : faFolder;

  return (
    <div className="Directory" onDoubleClick={doubleClicked}>
      <span className="icon" onClick={folderClicked}>
        <FontAwesomeIcon icon={icon} />
      </span>
      <span className="name">
        {name}
      </span>
    </div>
  );
};

Directory.propTypes = {
  name: PropTypes.string.isRequired,
  opened: PropTypes.bool.isRequired,
  doubleClicked: PropTypes.func.isRequired,
  folderClicked: PropTypes.func.isRequired,
};

export default Directory;
