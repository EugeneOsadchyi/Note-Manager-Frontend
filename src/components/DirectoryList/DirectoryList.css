ul.DirectoryList {
  margin: 5px 0;
  padding-left: 0;
  list-style: none;
  user-select: none;
}

ul.DirectoryList ul {
  margin: 0;
  padding-left: 20px;
}
