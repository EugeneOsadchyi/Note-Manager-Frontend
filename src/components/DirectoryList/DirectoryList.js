import React from 'react';
import PropTypes from 'prop-types';
import Directory from './Directory';

const DirectoryList = ({
  directories,
  parentId,
  toggleDirectory,
  editDirectory,
}) => (
  <ul>
    {
      directories.map((directory) => {
        if (directory.parentId === parentId) {
          return (
            <li key={directory.id}>
              <Directory
                name={directory.name}
                opened={directory.opened}
                folderClicked={toggleDirectory}
                doubleClicked={editDirectory}
              />
              {
                directory.opened ? (
                  <DirectoryList
                    directories={directories}
                    toggleDirectory={toggleDirectory}
                    editDirectory={editDirectory}
                    parentId={directory.id}
                  />
                ) : null
              }
            </li>
          );
        }
        return null;
      })
    }
  </ul>
);

DirectoryList.propTypes = {
  directories: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      opened: PropTypes.bool.isRequired,
    }),
  ),
  parentId: PropTypes.number,
  toggleDirectory: PropTypes.func.isRequired,
  editDirectory: PropTypes.func.isRequired,
};

DirectoryList.defaultProps = {
  directories: [],
  parentId: undefined,
};

export default DirectoryList;
