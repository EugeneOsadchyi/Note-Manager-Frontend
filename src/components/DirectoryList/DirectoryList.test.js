import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import DirectoryList from './DirectoryList';

describe('DirectoryList', () => {
  const mockSelectDirectory = jest.fn();
  const mockToggleDirectory = jest.fn();
  const mockEditDirectory = jest.fn();

  const mockedFunctions = {
    selectDirectory: mockSelectDirectory,
    toggleDirectory: mockToggleDirectory,
    editDirectory: mockEditDirectory,
  };

  describe('when `directories` are not passed', () => {
    const props = {
      directories: [],
      ...mockedFunctions,
    };

    const directoryList = shallow(<DirectoryList {...props} />);

    it('renders properly', () => {
      expect(shallowToJson(directoryList)).toMatchSnapshot();
    });

    it('does not contain Directory component', () => {
      expect(directoryList.find('Directory').exists()).toBe(false);
    });
  });

  describe('when component is fulfilled with `directories`', () => {
    const directories = [{
      id: 1,
      name: 'Root',
      opened: true,
      active: true,
    }];

    const props = {
      directories,
      ...mockedFunctions,
    };

    const directoryList = shallow(<DirectoryList {...props} />);

    it('renders properly', () => {
      expect(shallowToJson(directoryList)).toMatchSnapshot();
    });

    it('contains Directory component', () => {
      expect(directoryList.find('Directory').exists()).toBe(true);
    });

    describe('when directory is opened', () => {
      it('contains DirectoryList component', () => {
        expect(directoryList.find('DirectoryList').exists()).toBe(true);
      });
    });
  });
});
