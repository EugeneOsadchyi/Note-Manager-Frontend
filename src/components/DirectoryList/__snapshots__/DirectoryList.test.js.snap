// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectoryList when \`directories\` are not passed renders properly 1`] = `<ul />`;

exports[`DirectoryList when component is fulfilled with \`directories\` renders properly 1`] = `
<ul>
  <li
    key="1"
  >
    <Directory
      doubleClicked={[Function]}
      folderClicked={[Function]}
      name="Root"
      opened={true}
    />
    <DirectoryList
      directories={
        Array [
          Object {
            "id": 1,
            "name": "Root",
            "opened": true,
          },
        ]
      }
      editDirectory={[Function]}
      parentId={1}
      toggleDirectory={[Function]}
    />
  </li>
</ul>
`;
